prov/shm/src/smr_ep.c

```diff
@@ -319,6 +319,86 @@ void smr_format_iov(struct smr_cmd *cmd, fi_addr_t peer_id,
 	smr_post_pend_resp(pend, cmd, context, iov, count, resp);
 }
 
+int smr_mmap_copy_format(struct smr_ep *ep, struct smr_region *peer_smr,
+		    struct smr_cmd *cmd, fi_addr_t peer_id,
+		    const struct iovec *iov, size_t count, size_t total_len,
+		    uint32_t op, uint64_t tag, uint64_t data, uint64_t op_flags,
+		    void *context, struct smr_resp *resp,
+		    struct smr_tx_entry *pend)
+{
+	char shm_name[NAME_MAX];
+	void *mapped_ptr;
+	int fd, ret, num;
+	uint64_t msg_id;
+	struct smr_ep_name *map_name;
+
+	msg_id = ep->msg_id++;
+	num = snprintf(shm_name, NAME_MAX - 1, "%s_%s_%ld", ep->name,
+		       (char *)peer_smr + peer_smr->name_offset, msg_id);
+	if (num < 0) {
+		FI_WARN(&smr_prov, FI_LOG_AV, "generating shm file name failed\n");
+		return -errno;
+	}
+
+	fd = shm_open(shm_name, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+	if (fd < 0) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "shm_open error\n");
+		return -errno;
+	}
+
+	map_name = calloc(1, sizeof(struct smr_ep_name));
+	if (!map_name) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "calloc error\n");
+		ret = -FI_ENOMEM;
+		goto unlink_close;
+	}
+	strncpy((map_name)->name, shm_name, NAME_MAX - 1);
+	dlist_insert_tail(&(map_name)->entry, &ep_name_list);
+
+	ret = ftruncate(fd, total_len);
+	if (ret < 0) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "ftruncate error\n");
+		goto remove_entry;
+	}
+
+	mapped_ptr = mmap(NULL, total_len, PROT_READ | PROT_WRITE,
+			  MAP_SHARED, fd, 0);
+	if (mapped_ptr == MAP_FAILED) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "mmap error\n");
+		ret = -errno;
+		goto remove_entry;
+	}
+
+	num = ofi_copy_from_iov(mapped_ptr, total_len, iov, count, 0);
+	if (num != total_len) {
+		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "copy from iov error\n");
+		ret = -FI_EIO;
+		goto munmap;
+	}
```

Comment:
I don't love the name num for a size. Plus we can remove it entirely with
```if (ofi_copy_from_iov(mapped_ptr, total_len, iov, count, 0) != total_len)```

Reply:
Done.